### ISSUE TITLE
Significant performance degradation from Exllamav2 0.12 update

### ISSUE THREAD

**Issue by aikitoria** [oobabooga/text-generation-webui#5383]
### Describe the bug

I'm not entirely sure if this bug is caused by ooba or exllamav2 itself, but ever since the commit from 2 days ago that upgraded Exllamav2, my generations are significantly slower.

Previously, Goliath 120B 4.85bpw would produce around 14-15 tokens/s on an A100 80GB.
Now, it can only 10-11 tokens/s. The initial context processing is also much slower.

There's a comparable reduction in performance on H100 GPUs.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

1) Load a large exl2 model, i.e. https://huggingface.co/Panchovix/goliath-120b-exl2-rpcal/tree/4.85bpw
2) Generate some text with the current version and the one from 4 days ago at least two times, watch the token/s output.

### Logs

```shell
No errors, however, this warning message prints before the first generation.

/workspace/text-generation-webui/installer_files/env/lib/python3.11/site-packages/transformers/generation/configuration_utils.py:452: UserWarning: `num_beams` is set to 1. However, `length_penalty` is set to `0.0` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `length_penalty`.
  warnings.warn(
```

### System Info

```shell
RunPod A100 SXM server.
Runpod H100 SXM server.
```

**Comment by bartowski1182** [comment]
can you confirm your settings for the GPU split? is it possible you're accidentally putting more on the a100 now? it's a good bit slower than the h100 so if you're accidentally using it more then that would change performance enough like this

**Comment by aikitoria** [comment]
I'm testing on servers with one GPU only, so there's no split configured. All generation settings and model are exactly the same between the runs, the only thing I am changing is whether I am using the current main branch or the state before the last merge ( which I have captured here: https://github.com/aikitoria/text-generation-webui )

It's > 30% slower on either GPU after the upgrade.

**Comment by bartowski1182** [comment]
oh, i'm surprised a single H100 can run goliath at 4.25 bpw, interesting. 

On latest for a 7b model with my 3090, i get 60 tps

on previous snapshot I get 59.49 tps

So i don't know if it's actually an exllamav2 or text-generation-webui issue

**Comment by aikitoria** [comment]
It could be that the issue disproportionately affects huge models like Goliath, without having any noticeable difference on 7B

**Comment by bartowski1182** [comment]
yeah that second part of your comment is a known thing, it's like that with most quant types, it's like a "warmup" stage that it's in, and why for benchmarks there's typically a warmup stage or you discard the first few results

**Comment by aikitoria** [comment]
A single A100/H100 can run goliath indeed, even with long context, like these quants https://huggingface.co/aikitoria/Goliath-longLORA-120b-rope8-32k-exl2

Was testing these earlier and wondering why the performance suddently cratered until I remembered I had reinstalled ooba. If I downgrade it and reinstall all the deps then the performance returns.

**Comment by bartowski1182** [comment]
regarding disproportionally affecting larger models, that's definitely a reasonable theory, i'll test between snapshot and latest for a 34b model, think you could try the same? or is it a hassle for you to switch between them

**Comment by aikitoria** [comment]
Good idea, let me try to load a 34b with each on the A100

**Comment by bartowski1182** [comment]
latest: Output generated in 13.43 seconds (16.60 tokens/s, 223 tokens, context 154, seed 203521508)

snapshot-2024-01-14: Output generated in 12.39 seconds (17.27 tokens/s, 214 tokens, context 154, seed 1608668989)

So that is a 4% difference, enough to possibly give validity to the theory that the bigger the model the more it's affected

**Comment by aikitoria** [comment]
I tried running LoneStriker/Nous-Hermes-2-Yi-34B-8.0bpw-h8-exl2 on both versions.

Old ver:
```
Output generated in 8.12 seconds (21.44 tokens/s, 174 tokens, context 3418, seed 608977201)
Output generated in 8.50 seconds (30.00 tokens/s, 255 tokens, context 3418, seed 1831287466)
Output generated in 4.40 seconds (28.88 tokens/s, 127 tokens, context 3418, seed 238612891)
Output generated in 4.49 seconds (28.28 tokens/s, 127 tokens, context 3418, seed 440243725)
```

New ver: 
```
Output generated in 7.96 seconds (22.74 tokens/s, 181 tokens, context 3418, seed 1462665345)
Output generated in 7.90 seconds (29.48 tokens/s, 233 tokens, context 3418, seed 1316551388)
Output generated in 6.13 seconds (29.35 tokens/s, 180 tokens, context 3418, seed 423543194)
Output generated in 8.78 seconds (29.03 tokens/s, 255 tokens, context 3418, seed 320926810)
```

Not conclusive. Seems at 34b there isn't a difference. Now to try a 70b Llama 2 model like Goliath is based on.

**Comment by Ph0rk0z** [comment]
I've been git pulling exllama and didn't see any difference in speed on ampere cards. I run 2x24g and usually models like 70b+

**Comment by aikitoria** [comment]
I tried running grimulkan/Euryale-1.3-longLORA-70b-rope8-32k-6bpw_h8_exl2 on both versions.

Old ver:
```
Output generated in 10.39 seconds (13.86 tokens/s, 144 tokens, context 3697, seed 1744151427)
Output generated in 12.22 seconds (20.87 tokens/s, 255 tokens, context 3697, seed 2053525482)
Output generated in 4.78 seconds (20.08 tokens/s, 96 tokens, context 3697, seed 1730654491)
Output generated in 9.56 seconds (21.12 tokens/s, 202 tokens, context 3697, seed 867148768)
```
New ver:
```
Output generated in 16.66 seconds (15.30 tokens/s, 255 tokens, context 3697, seed 1864681781)
Output generated in 6.98 seconds (18.48 tokens/s, 129 tokens, context 3697, seed 1075040978)
Output generated in 5.33 seconds (18.75 tokens/s, 100 tokens, context 3697, seed 77696659)
Output generated in 12.49 seconds (19.29 tokens/s, 241 tokens, context 3697, seed 1783216709
```
There is a measurable performance degradation here for 70B that seems too large to be within margin of error. It's not as bad as Goliath, but it's there.

**Comment by Ph0rk0z** [comment]
Output a fixed number of tokens. Turn off EOS and set an amount like 150 or something. That will tell you exactly how much it fell by.

**Comment by aikitoria** [comment]
I guess that would be a better test indeed, but isn't looking at the tokens/s number at ~similar length outputs good enough from the second generation onwards?

**Comment by aikitoria** [comment]
#### For turboderp/Mistral-7B-instruct-exl2:6.0bpw
Old ver:
```
Output generated in 3.08 seconds (82.73 tokens/s, 255 tokens, context 3506, seed 2066395078)
Output generated in 3.07 seconds (83.12 tokens/s, 255 tokens, context 3506, seed 1696949787)
Output generated in 3.08 seconds (82.83 tokens/s, 255 tokens, context 3506, seed 2032979249)
Output generated in 3.09 seconds (82.60 tokens/s, 255 tokens, context 3506, seed 1228333612)
```
New ver:
```
Output generated in 3.14 seconds (81.24 tokens/s, 255 tokens, context 3506, seed 847691391)
Output generated in 3.14 seconds (81.20 tokens/s, 255 tokens, context 3506, seed 2083584810)
Output generated in 3.81 seconds (66.95 tokens/s, 255 tokens, context 3506, seed 142448973)
Output generated in 3.58 seconds (71.22 tokens/s, 255 tokens, context 3506, seed 1427862816)
```
This one appears more inconsistent for some reason.

#### For LoneStriker/Nous-Hermes-2-Yi-34B-8.0bpw-h8-exl2
Old ver:
```
Output generated in 8.47 seconds (30.11 tokens/s, 255 tokens, context 3418, seed 559987598)
Output generated in 8.60 seconds (29.64 tokens/s, 255 tokens, context 3418, seed 614006132)
Output generated in 8.64 seconds (29.51 tokens/s, 255 tokens, context 3418, seed 300744856)
Output generated in 8.47 seconds (30.12 tokens/s, 255 tokens, context 3418, seed 1731314794)
```
New ver:
```
Output generated in 8.76 seconds (29.11 tokens/s, 255 tokens, context 3418, seed 353808507)
Output generated in 8.72 seconds (29.23 tokens/s, 255 tokens, context 3418, seed 355136799)
Output generated in 8.70 seconds (29.32 tokens/s, 255 tokens, context 3418, seed 1007109035)
Output generated in 9.06 seconds (28.13 tokens/s, 255 tokens, context 3418, seed 279263108)
```
This one is very slightly slower.

#### For grimulkan/Euryale-1.3-longLORA-70b-rope8-32k-6bpw_h8_exl2
Old ver:
```
Output generated in 12.24 seconds (20.84 tokens/s, 255 tokens, context 3697, seed 2020886887)
Output generated in 12.22 seconds (20.87 tokens/s, 255 tokens, context 3697, seed 898298057)
Output generated in 12.21 seconds (20.89 tokens/s, 255 tokens, context 3697, seed 130012350)
Output generated in 12.23 seconds (20.84 tokens/s, 255 tokens, context 3697, seed 336479399)
```
New ver:
```
Output generated in 13.20 seconds (19.31 tokens/s, 255 tokens, context 3697, seed 1445970018)
Output generated in 13.23 seconds (19.28 tokens/s, 255 tokens, context 3697, seed 728081663)
Output generated in 13.23 seconds (19.27 tokens/s, 255 tokens, context 3697, seed 302005717)
Output generated in 13.22 seconds (19.29 tokens/s, 255 tokens, context 3697, seed 1288134348)
```
This one is slower.

#### For Panchovix/goliath-120b-exl2-rpcal:4.85bpw
Old ver:
```
Output generated in 17.65 seconds (14.45 tokens/s, 255 tokens, context 3697, seed 1249076939)
Output generated in 17.64 seconds (14.45 tokens/s, 255 tokens, context 3697, seed 944567877)
Output generated in 17.57 seconds (14.51 tokens/s, 255 tokens, context 3697, seed 1576820871)
Output generated in 17.60 seconds (14.49 tokens/s, 255 tokens, context 3697, seed 366032215)
```
New ver:
```
Output generated in 21.99 seconds (11.60 tokens/s, 255 tokens, context 3697, seed 1089410131)
Output generated in 21.95 seconds (11.62 tokens/s, 255 tokens, context 3697, seed 802950688)
Output generated in 21.95 seconds (11.62 tokens/s, 255 tokens, context 3697, seed 57410179)
Output generated in 21.99 seconds (11.60 tokens/s, 255 tokens, context 3697, seed 1911223589)
```
This one is much slower.

All tests run on the exact same server with A100 80GB. I skipped the first generation for each run.

**Comment by aikitoria** [comment]
Reading the changelog for exllamav2 0.12, these things seem potentially interesting:
- Add freq/pres penalties
- Disable flash-attn for non-causal attention (fixes left-padding until FA2 implements custom bias)

I thought the cause might be the frequency penalty, however even if I remove that from my generation settings it does not change the performance.

I don't know what non-causal attention is. Does that apply to these models?

**Comment by Ph0rk0z** [comment]
Disabling flash attention could cause speed drops but these are all causal llm.

Perhaps this is an issue: https://github.com/turboderp/exllamav2/commit/7a9d12ae4c9611c3e6379d43ad536b52a1f97618

Seems like a big rewrite to RMS norms.

Pull right before that commit and compile then test.

**Comment by aikitoria** [comment]
Alright, I will try this tomorrow

**Comment by Ph0rk0z** [comment]
That does a lot of stuff. I just go into my environment and do `python server.py` to start it.

You should probably switch into the same environment when compiling exllama as well.

**Comment by bartowski1182** [comment]
i suppose the truly easiest thing to do would be to run exllamav2 directly rather than through ooba, remove variables too at the same time in case it's somehow unrelated to the bump (unlikely)

**Comment by aikitoria** [comment]
> That does a lot of stuff. I just go into my environment and do `python server.py` to start it.
> 
> You should probably switch into the same environment when compiling exllama as well.

Leaving the steps here for how I did this in case someone else needs it.
Did I do it in a particularly clunky way or is this how it's meant to be done?

1) Activate the environment made by the installer:
`source installer_files/conda/etc/profile.d/conda.sh`
`conda activate installer_files/env/`
2) Get rid of current exllamav2 build (this step also needs to be repeated before every compile):
`pip uninstall exllamav2`
3) Get the new exllamav2:
`git clone https://github.com/turboderp/exllamav2`
4) Build it:
`cd exllamav2`
`python setup.py install --user`
`cd ..`
5) Start webui:
`python server.py`

Now I can test the performance for different commits easily. Using only 128 tokens this time.

`tags/v0.0.11`:
```
Output generated in 8.78 seconds (14.47 tokens/s, 127 tokens, context 1682, seed 2142719412)
Output generated in 8.84 seconds (14.37 tokens/s, 127 tokens, context 1682, seed 131042970)
```
`9373d0` (one before the commit you mentioned):
```
Output generated in 11.11 seconds (11.43 tokens/s, 127 tokens, context 1682, seed 295294536)
Output generated in 11.12 seconds (11.42 tokens/s, 127 tokens, context 1682, seed 245566471)
```
`master`:
```
Output generated in 11.12 seconds (11.42 tokens/s, 127 tokens, context 1682, seed 913349047)
Output generated in 11.21 seconds (11.33 tokens/s, 127 tokens, context 1682, seed 921038948)
```

Looks like that wasn't it. I'll do binary search to find the problem...

**Comment by aikitoria** [comment]
This commit is the problem: https://github.com/turboderp/exllamav2/commit/87225fe0

**Comment by aikitoria** [comment]
I tried changing those parameters back to what they were before:

`exllamav2/exllamav2_ext/config.h`:
```
#define MAX_Q_GEMM_ROWS 50
#define MAX_Q_GEMM_ROWS_KERNEL 6
```

But that does not fix the performance.

**Comment by aikitoria** [comment]
I wonder if the problem is with the dynamic dispatch that was added to `pick_gemm_half_q_half_kernel` or `pick_gemm_half_q_half_kernel` or something else. So many changes at once, hard to tell...

**Comment by aikitoria** [comment]
I just to make sure nothing strange is happening because of ooba, I also tried using the test_inference.py that comes with it.

Before: `Response generated in 8.22 seconds, 128 tokens, 15.57 tokens/second (includes prompt eval.)`
After:  `Response generated in 10.41 seconds, 128 tokens, 12.29 tokens/second (includes prompt eval.)`

**Comment by Ph0rk0z** [comment]
Makes sense. Lllama.cpp did similar. Better batch often hurts single perf.

**Comment by aikitoria** [comment]
Why? Couldn't it compile separate kernels for batch and single mode?

**Comment by Ph0rk0z** [comment]
That's mostly what they end up doing.

**Comment by aikitoria** [comment]
This issue has been fixed in exllamav2 0.0.13, so all will be good once this project updates to that

**Comment by aikitoria** [comment]
Closing this as it is fixed on dev branch
